src/backend/mpi/swdsm.cpp

```diff
@@ -436,315 +433,262 @@ void handler(int sig, siginfo_t *si, void *unused){
-	int n;
-	homenode = getHomenode(lineAddr);
-
-	if(prevsharer==0 ){ //if there is strictly less than two 'stable' sharers
-		MPI_Win_lock(MPI_LOCK_SHARED, homenode, 0, sharerWindow);
-		MPI_Get_accumulate(&id, 1, MPI_LONG, &tempsharer, 1, MPI_LONG,
-			homenode, classidx, 1, MPI_LONG, MPI_BOR, sharerWindow);
-		MPI_Get(&tempwriter, 1,MPI_LONG,homenode,classidx+1,1,MPI_LONG,sharerWindow);
-		MPI_Win_unlock(homenode, sharerWindow);
-	}
-
-	MPI_Win_lock(MPI_LOCK_EXCLUSIVE, workrank, 0, sharerWindow);
-	globalSharers[classidx] |= tempsharer;
-	globalSharers[classidx+1] |= tempwriter;
-	MPI_Win_unlock(workrank, sharerWindow);
-
-	unsigned long offset = getOffset(lineAddr);
-	if(isPowerOf2((tempsharer)&invid) && tempsharer != id && prevsharer == 0){ //Other private. but may not have loaded page yet.
-		unsigned long ownid = tempsharer&invid; // remove own bit
-		unsigned long owner = invalid_node; // initialize to failsafe value
-		for(n=0; n<numtasks; n++) {
-			if(1ul<<n==ownid) {
-				owner = n; //just get rank...
-				break;
+	for(std::size_t i = 0; i < fetch_size; i+=CACHELINE){
+		if(pages_to_load[i]){
+			/* Check local pyxis directory if we are sharer of the page */
+			local_sharers[i] = (globalSharers[classification_index_array[i]])&node_id_bit;
+			if(local_sharers[i] == 0){
+				sharer_bit_mask[i*2] = node_id_bit;
+				new_sharer = true; //At least one new sharer detected
 			}
 		}
-		if(owner != invalid_node) {
-			MPI_Win_lock(MPI_LOCK_EXCLUSIVE, owner, 0, sharerWindow);
-			MPI_Accumulate(&id, 1, MPI_LONG, owner, classidx, 1, MPI_LONG, MPI_BOR, sharerWindow);
-			MPI_Win_unlock(owner, sharerWindow);
-		}
-
 	}
+	MPI_Win_unlock(workrank, sharerWindow);
 
-	MPI_Win_lock(MPI_LOCK_SHARED, homenode , 0, globalDataWindow[homenode]);
-	MPI_Get(&cacheData[startidx*pagesize],
-					1,
-					cacheblock,
-					homenode,
-					offset, 1,cacheblock,globalDataWindow[homenode]);
-	MPI_Win_unlock(homenode, globalDataWindow[homenode]);
-
-	if(cacheControl[startidx].tag == GLOBAL_NULL){
-		vm::map_memory(lineptr, blocksize, pagesize*startidx, PROT_READ);
-		cacheControl[startidx].tag = lineAddr;
-	}
-	else{
-		mprotect(lineptr,pagesize*CACHELINE,PROT_READ);
-	}
-	touchedcache[startidx] = 1;
-	cacheControl[startidx].state = VALID;
-
-	cacheControl[startidx].dirty=CLEAN;
-	sem_post(&ibsem);
-}
-
-void prefetch_cache_entry(unsigned long prefetchtag, unsigned long prefetchline) {
-	int i;
-	unsigned long homenode;
-	unsigned long id = 1 << getID();
-	unsigned long invid = ~id;
-	if(prefetchtag>=size_of_all){//Trying to access/prefetch out of memory
-		return;
-	}
-
-
-	homenode = getHomenode(prefetchtag);
-	unsigned long cacheIndex = prefetchline;
-	if(cacheIndex >= cachesize){
-		printf("idx > size   cacheIndex:%ld cachesize:%ld\n",cacheIndex,cachesize);
-		return;
+	/* If this node is a new sharer of at least one of the pages */
+	if(new_sharer){
+		/* Register ourselves on all newly shared pages the loadnode directory */
```

Comment:
Specify who "ourselves" is.  Also loadnode -> load_node.

Reply:
I pushed a more elaborate comment including the changes requested, please let me know if it is clear enough.